### PR TITLE
[Feature] Fixed sampler with limit on sampled nodes/edges in batch subgraph

### PR DESCRIPTION
**Pull request by ayushnoori**
## Description
Subgraph sampler that sets an upper bound on the number of nodes included in each layer of the sampled subgraph. At each layer, the frontier is randomly subsampled. Rare node types can also be upsampled by taking the scaled square root of the sampling probabilities (best strategy TBD). The new `FixedSampler` performs node-wise neighbor sampling and returns the subgraph induced by all the sampled nodes.

The relevant issue is: https://github.com/dmlc/dgl/issues/6623, thanks to @frozenbugs and @jermainewang for their input and review.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] Code is well-documented
- [X] Related issue is referred in this PR
- [ ] All changes have test coverage

Please note that unit tests for `FixedSampler` are not yet written.

### CODE REVIEW

File: python/dgl/dataloading/fixed.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    It performs node-wise neighbor sampling and returns the subgraph induced by
+    all the sampled nodes.
+
+    Parameters
+    ----------
+    fanouts : list[int] or list[dict[etype, int]]
+        List of neighbors to sample per edge type for each GNN layer, with the i-th
+        element being the fanout for the i-th GNN layer.
+
+        If only a single integer is provided, DGL assumes that every edge type
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:26]: Can you follow this example to arrange the bullet points? Please pay attention to the indent.
https://github.com/dmlc/dgl/pull/6998/files#diff-d94a88f81d4e382b3c481a2e2e1b7383ac0b61e39e0d364e8cf457d627356ab8R33

File: python/dgl/dataloading/fixed.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        will have the same fanout.
+
+        If -1 is provided for one edge type on one layer, then all inbound edges
+        of that edge type will be included.
+    fixed_k : int
+            The number of nodes to sample for each GNN layer.
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:32]: nit: indent

File: python/dgl/dataloading/fixed.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+import torch
+from collections import defaultdict
+
+class FixedSampler(Sampler):
+    """Subgraph sampler that sets an upper bound on the number of nodes included in 
+    each layer of the sampled subgraph.
+    
+    At each layer, the frontier is randomly subsampled. Rare node types can also be 
+    upsampled by taking the scaled square root of the sampling probabilities.
+
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:16]: nit: arrange the doc to 1 paragraph.

File: python/dgl/dataloading/fixed.py
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:1]: fixed_sampler.py / capped_neighbor_sampler.py

File: python/dgl/dataloading/fixed.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
+"""Fixed subgraph sampler."""
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:1]: please take action to fix the lint error.

File: python/dgl/dataloading/fixed.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+from ..sampling.utils import EidExcluder
+from .base import set_node_lazy_features, set_edge_lazy_features, Sampler
+
+# import non-DGL libraries
+import numpy as np
+import torch
+from collections import defaultdict
+
+class FixedSampler(Sampler):
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:10]: Not sure whether FixedSampler is a good name or not, maybe CappedNeighborSampler?

File: python/dgl/dataloading/fixed.py (hunk 3 above)
Comment (frozenbugs) [reply python/dgl/dataloading/fixed.py:16]: try to use ChatGPT or Bard to help you polish the wording.

File: python/dgl/dataloading/fixed.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        self.prefetch_edge_feats = prefetch_edge_feats
+        self.output_device = output_device
+
+    def sample(self, g, seed_nodes, exclude_eids=None):
+        """Sampling function.
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph to sampler from.
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:60]:
```suggestion
            The graph to sample from.
```

File: python/dgl/dataloading/fixed.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        seed_nodes : Tensor or dict[str, Tensor]
+            The nodes sampled in the current minibatch.
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:62]: This is not very accurate.

File: python/dgl/dataloading/fixed.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        exclude_eids : Tensor or dict[etype, Tensor], optional
+            The edges to exclude from neighborhood expansion.
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:64]: Edges to exclude from the sampled subgraph.

File: python/dgl/dataloading/fixed.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+        Returns
+        -------
+        input_nodes, output_nodes, subg
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:68]: What are the types? Can you arrange them as Parameters?

File: python/dgl/dataloading/fixed.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            A triplet containing (1) the node IDs inducing the subgraph, (2) the node
+            IDs that are sampled in this minibatch, and (3) the subgraph itself.
+        """
+
+        # define empty dictionary to store reached nodes
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:73]:
```suggestion
        # Define empty dictionary to store reached nodes.
```
nit: Same for other comments, treat it as an English sentence.

File: python/dgl/dataloading/fixed.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        output_nodes = seed_nodes
+        all_reached_nodes = [seed_nodes]
+
+        # iterate over fanout
+        for fanout in reversed(self.fanouts):
+
+            # sample frontier
+            frontier = g.sample_neighbors(
+                seed_nodes, fanout, output_device=self.output_device,
+                replace=self.replace, prob=self.prob, exclude_edges=exclude_eids)
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:83]: max 80 chars.

File: python/dgl/dataloading/fixed.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+            # get reached nodes
+            curr_reached = defaultdict(list)
+            for c_etype in frontier.canonical_etypes:
+                (src_type, rel_type, dst_type) = c_etype
+                src, _ = frontier.edges(etype = c_etype)
+                curr_reached[src_type].append(src)
+
+            # de-duplication
+            curr_reached = {ntype : torch.unique(torch.cat(srcs)) for ntype, srcs in curr_reached.items()}
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:93]: max 80 chars.

File: python/dgl/dataloading/fixed.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+            merged_nodes[ntype] = torch.unique(torch.cat([reached.get(ntype, []) for reached in all_reached_nodes]))
+        subg = g.subgraph(merged_nodes, relabel_nodes=True, output_device=self.output_device)
+
+        if exclude_eids is not None:
+            subg = EidExcluder(exclude_eids)(subg)
+
+        set_node_lazy_features(subg, self.prefetch_node_feats)
+        set_edge_lazy_features(subg, self.prefetch_edge_feats)
+
+        return seed_nodes, output_nodes, subg
```
Comment (frozenbugs) [inline python/dgl/dataloading/fixed.py:146]: add a newline to the end of the file.

**Comment by ayushnoori** [comment]
Thanks for the edits, @frozenbugs. Happy to address all the formatting errors. Do you see any issues with the function implementation itself?

File: python/dgl/dataloading/fixed.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
+"""Fixed subgraph sampler."""
+from ..sampling.utils import EidExcluder
+from .base import set_node_lazy_features, set_edge_lazy_features, Sampler
+
+# import non-DGL libraries
```
[the remark this answers is not in the document]
Comment (ayushnoori) [reply python/dgl/dataloading/fixed.py:5]: Removed.

File: python/dgl/dataloading/fixed.py (hunk 5 above)
Comment (ayushnoori) [reply python/dgl/dataloading/fixed.py:10]: Updated to `CappedNeighborSampler`, can also change the file name.

File: python/dgl/dataloading/fixed.py (hunk 1 above)
Comment (ayushnoori) [reply python/dgl/dataloading/fixed.py:26]: Updated to match `python/dgl/graphbolt/sampled_subgraph.py`.

File: python/dgl/dataloading/fixed.py (hunk 6 above)
Comment (ayushnoori) [reply python/dgl/dataloading/fixed.py:60]: Not sure what you mean by this? Doesn't `exclude_eids` define edges to exclude?

File: python/dgl/dataloading/fixed.py (hunk 7 above)
Comment (ayushnoori) [reply python/dgl/dataloading/fixed.py:62]: Currently, the description of `seed_nodes` is, "The nodes sampled in the current minibatch." How would you suggest we change it?

File: python/dgl/dataloading/fixed.py (hunk 10 above)
Comment (ayushnoori) [reply python/dgl/dataloading/fixed.py:73]: Updated for all comments.

File: python/dgl/dataloading/fixed.py (hunk 7 above)
Comment (ayushnoori) [reply python/dgl/dataloading/fixed.py:62]: How about "Nodes which induce the subgraph." - would that work?

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
@dgl-bot

**Comment by frozenbugs** [comment]
```
************* Module dgl.dataloading.capped_neighbor_sampler
python/dgl/dataloading/capped_neighbor_sampler.py:60:4: W0221: Parameters differ from overridden 'sample' method (arguments-differ)
python/dgl/dataloading/capped_neighbor_sampler.py:151:27: C0103: Variable name "node_IDs" doesn't conform to snake_case naming style (invalid-name)
python/dgl/dataloading/capped_neighbor_sampler.py:102:27: W0612: Unused variable 'rel_type' (unused-variable)
python/dgl/dataloading/capped_neighbor_sampler.py:102:37: W0612: Unused variable 'dst_type' (unused-variable)
```

**Comment by frozenbugs** [comment]
@ayushnoori if you don't plan to write a unit test, we can also merge this, can you add a comment in capped_neighbor_sampler.py, says "This code was contributed by a community member (ayushnoori). There aren't currently any unit tests in place to verify its functionality, so please be cautious if you need to make any changes to the code's logic."

**Comment by mfbalin** [comment]
@dgl-bot

**Comment by ayushnoori** [comment]
Hi @frozenbugs, I'm kindly following up with this PR. I've added the requested comment – please let me know if this LGTM. Thanks again!

**Comment by ayushnoori** [comment]
Hmm, not sure why CI test failed. I already ran `lintrunner -a`, this is what I see:
```
> lintrunner -a
Warning: Could not find a lintrunner config at: '.lintrunner.private.toml'. Continuing without using configuration file.
WARNING: No previous init data found. If this is the first time you're running lintrunner, you should run `lintrunner init`.
ok No lint issues.
Successfully applied all patches.

> git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

Have also merged changes so this PR is up-to-date.

**Comment by mfbalin** [comment]
https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-6668/10/10/logs/logs_dir/tmp1ptbor3n.log

```
************* Module dgl.dataloading.capped_neighbor_sampler
python/dgl/dataloading/capped_neighbor_sampler.py:65:4: W0221: Parameters differ from overridden 'sample' method (arguments-differ)
python/dgl/dataloading/capped_neighbor_sampler.py:156:27: C0103: Variable name "node_IDs" doesn't conform to snake_case naming style (invalid-name)
python/dgl/dataloading/capped_neighbor_sampler.py:107:27: W0612: Unused variable 'rel_type' (unused-variable)
python/dgl/dataloading/capped_neighbor_sampler.py:107:37: W0612: Unused variable 'dst_type' (unused-variable)
```

**Comment by ayushnoori** [comment]
Thanks, @mfbalin. Should be fixed now.

**Comment by mfbalin** [comment]
@dgl-bot

**Comment by ayushnoori** [comment]
Looks like the error still is:
```
************* Module dgl.dataloading.capped_neighbor_sampler
python/dgl/dataloading/capped_neighbor_sampler.py:65:4: W0221: Parameters differ from overridden 'sample' method (arguments-differ)
```
The function signature in my code is:
```python
def sample(self, g, indices, exclude_eids=None):
```
The `sample()` function signature of [`dgl.dataloading.base.Sampler`](https://docs.dgl.ai/en/latest/_modules/dgl/dataloading/base.html#Sampler) is:
```
def sample(self, g, indices):
```

This is intentional, I added an optional `exclude_eids` argument. All other parameters should match. @mfbalin, may you please advise why this is a problem? Thanks!

**Comment by mfbalin** [comment]
@ayushnoori You can add `  # pylint: disable=arguments-differ` to disable the warning if you really need it to be that way.

**Comment by ayushnoori** [comment]
Thanks, @mfbalin – done.

**Comment by mfbalin** [comment]
@dgl-bot

**Review by ayushnoori — COMMENTED** [review]
Feedback from @frozenbugs should now be addressed.

File: python/dgl/dataloading/capped_neighbor_sampler.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
+"""Fixed subgraph sampler."""
```
Comment (frozenbugs) [inline python/dgl/dataloading/capped_neighbor_sampler.py:1]:
```suggestion
"""Capped neighbor sampler."""
```

**Review by frozenbugs — APPROVED** [review]

**Comment by frozenbugs** [comment]
Hi @ayushnoori , I have this PR merged. And it can be used in nightly DGL wheel, and next DGL release.
But please be aware that the lagacy dataloading will be fully deprecated and removed as I mentioned before, so soon in the future the code will be removed together with the legacy dataloading. If you want to retain it in DGL, please follow up with a Graphbolt implementation.

**Comment by ayushnoori** [comment]
Understood, thanks so much, @frozenbugs!

**Pull request merged**